### ISSUE TITLE
Invalid domain name test in Subject Alternative Name

### ISSUE THREAD

**Issue by johndoe31415** [google/x509test#3]
Hi David,

first off, thanks for the great work with this, it has been invaluable for me. I'm currently writing a validator to detect all kinds of issues with certificates and your test corpus has been a goldmine.

In my implementation I have an issue with xf-ext-altname-invalid-domain.pem -- it might very well be that my understanding is wrong. I still open this up as a ticket because if it turns out to be bogus there'll be public documentation and others might find it before asking :-)

The certificate contains a CN of "\*.google.com" and a Subject Alternative Name of "DNS:*.123google.com". The SN/GN indicates:

SN = "RFC5280 s4.2.1.6 'The name MUST be in the preferred name syntax, as specified by Section 3.5 of [RFC1034] and as modified by Section 2.1 of [RFC1123]'"
GN = Subject Alternative Name with invalid domain

The issue here does not seem to be the mismatch between CN/SAN because if a SAN is present it always has precedence (to the best of my knowledge). Therefore the only issue I can see would be the wildcard domainname "*.123google.com". However, I don't see what's wrong with it. Leading digits for labels are okay (e.g., "www.321meins.de" leads to eBay), so maybe I'm missing something super obvious?

Would be very much appreciated if you could point me in the right direction.
Again, thank you so much for x509test.
Cheers,
Joe

**Comment by daviddrysdale** [comment]
Hmm, interesting.  I don't remember what my original intent was, so going back to the specs&hellip;

[RFC5280 s4.2.1.6](https://tools.ietf.org/html/rfc5280#section-4.2.1.6) has:

> The name MUST be in the "preferred name syntax", as specified by
> Section 3.5 of [RFC1034] and as modified by Section 2.1 of
> [RFC1123].

The [RFC1034](https://tools.ietf.org/html/rfc1034) s3.5 reference has a BNF definition that a) doesn't include `*` and b) requires the first character of a label to be a letter, not a digit.

The [RFC1123 s2.1](https://tools.ietf.org/html/rfc1123#section-2.1) update drops requirement b):

> the restriction on the first character is relaxed to allow either a
> letter or a digit 

but a) still appears to apply. 

That would mean that wildcard certs are invalid, so I must have missed something first time around.  Looking further&hellip;

RFC1034 s4.3.3 has discussion of wildcards, but only as an internal name server mechanism (s4.3); however, [RFC6125 s6.4.3](https://tools.ietf.org/html/rfc6125#section-6.4.3) does include discussion of `*` in DNS names (even as it tries to discourage their use in s7.2):

> A client employing this specification's rules MAY match the reference
> identifier against a presented identifier whose DNS domain name
> portion contains the wildcard character '*' as part or all of a label

So I think that means that the DNS name in that test case is valid in both theory and practice.

tl;dr: well-spotted, I'll change it to a different invalid value

**Comment by daviddrysdale** [comment]
I think `@` would work as an invalid character to replace the `*` with &ndash; @johndoe31415 do you agree?

**Comment by johndoe31415** [comment]
Hi @daviddrysdale -- super interesting. So RFC1034 mentions that valid labels are [-A-Za-z0-9]+ and the wildcard character "*" is a special case, expanded on in RFC6125. To the best of my knowledge, the "@" character is included in no grammar for labels, so I would think it would be safe to call it illegal for domain name labels. Thanks for doing so much digging!
